File: src/main/java/com/github/svegon/capi/event/network/S2CPlayPacketListener.java
package com.github.svegon.capi.event.network;

import com.github.svegon.capi.event.ListenerCollectionFactory;
import com.github.svegon.capi.event.ListenerList;
import com.github.svegon.capi.event.ListenerSet;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.network.listener.ClientPlayPacketListener;
import net.minecraft.network.packet.Packet;
import net.minecraft.network.packet.s2c.play.*;
import net.minecraft.text.Text;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

public interface S2CPlayPacketListener extends ClientPlayPacketListener {
    Event<S2CPlayPacketListener> CLIENT_PACKET_RECEIVED_EVENT = EventFactory.createArrayBacked(
            S2CPlayPacketListener.class,
            (listeners) -> new S2CPlayPacketListener() {

                @Override
                public void apply(Packet<ClientPlayPacketListener> packet, CallbackInfo info) {
                    for (S2CPlayPacketListener listener : listeners) {
                        listener.apply(packet, info);

                        if (info.isCancelled()) {
                            return;
                        }
                    }
                }
            });
    ListenerList<S2CPlayPacketListener> LISTENER_LIST = ListenerCollectionFactory.listenersVector(
            (l) -> new S2CPlayPacketListener() {
                @Override
                public void apply(Packet<ClientPlayPacketListener> packet, CallbackInfo info) {
                    for (S2CPlayPacketListener listener : l) {
                        listener.apply(packet, info);

                        if (info.isCancelled()) {
                            return;
                        }
                    }
                }
            });
    ListenerSet<S2CPlayPacketListener> LISTENER_SET = ListenerCollectionFactory.listenersLinkedSet(
            (s) -> new S2CPlayPacketListener() {
                @Override
                public void apply(Packet<ClientPlayPacketListener> packet, CallbackInfo info) {
                    for (S2CPlayPacketListener listener : s) {
                        listener.apply(packet, info);

                        if (info.isCancelled()) {
                            return;
                        }
                    }
                }
            });

    default void apply(Packet<ClientPlayPacketListener> packet, CallbackInfo info) {
        packet.apply(this);
    }

    @Override
    default boolean isConnectionOpen() {
        return false;
    }

    @Override
    default void onDamageTilt(DamageTiltS2CPacket packet) {

    }

    @Override
    default void onProfilelessChatMessage(ProfilelessChatMessageS2CPacket packet) {

    }

    @Override
    default void onRemoveMessage(RemoveMessageS2CPacket packet) {

    }

    @Override
    default void onChunkBiomeData(ChunkBiomeDataS2CPacket packet) {

    }

    @Override
    default void onPlayerRemove(PlayerRemoveS2CPacket packet) {

    }

    @Override
    default void onFeatures(FeaturesS2CPacket packet) {

    }

    @Override
    default void onBundle(BundleS2CPacket packet) {

    }

    @Override
    default void onEntityDamage(EntityDamageS2CPacket packet) {

    }
    @Override
    default void onEntitySpawn(EntitySpawnS2CPacket packet) {

    }

    @Override
    default void onExperienceOrbSpawn(ExperienceOrbSpawnS2CPacket packet) {

    }

    @Override
    default void onScoreboardObjectiveUpdate(ScoreboardObjectiveUpdateS2CPacket packet) {

    }

    @Override
    default void onPlayerSpawn(PlayerSpawnS2CPacket packet) {

    }

    @Override
    default void onEntityAnimation(EntityAnimationS2CPacket packet) {

    }

    @Override
    default void onStatistics(StatisticsS2CPacket packet) {

    }

    @Override
    default void onUnlockRecipes(UnlockRecipesS2CPacket packet) {

    }

    @Override
    default void onBlockBreakingProgress(BlockBreakingProgressS2CPacket packet) {

    }

    @Override
    default void onSignEditorOpen(SignEditorOpenS2CPacket packet) {

    }

    @Override
    default void onBlockEntityUpdate(BlockEntityUpdateS2CPacket packet) {

    }

    @Override
    default void onBlockEvent(BlockEventS2CPacket packet) {

    }

    @Override
    default void onBlockUpdate(BlockUpdateS2CPacket packet) {

    }

    @Override
    default void onGameMessage(GameMessageS2CPacket packet) {

    }

    @Override
    default void onChatMessage(ChatMessageS2CPacket packet) {

    }

    @Override
    default void onChunkDeltaUpdate(ChunkDeltaUpdateS2CPacket packet) {

    }

    @Override
    default void onMapUpdate(MapUpdateS2CPacket packet) {

    }

    @Override
    default void onCloseScreen(CloseScreenS2CPacket packet) {

    }

    @Override
    default void onInventory(InventoryS2CPacket packet) {

    }

    @Override
    default void onOpenHorseScreen(OpenHorseScreenS2CPacket packet) {

    }

    @Override
    default void onScreenHandlerPropertyUpdate(ScreenHandlerPropertyUpdateS2CPacket packet) {

    }

    @Override
    default void onScreenHandlerSlotUpdate(ScreenHandlerSlotUpdateS2CPacket packet) {

    }

    @Override
    default void onCustomPayload(CustomPayloadS2CPacket packet) {

    }

    @Override
    default void onDisconnect(DisconnectS2CPacket packet) {

    }

    @Override
    default void onEntityStatus(EntityStatusS2CPacket packet) {

    }

    @Override
    default void onEntityAttach(EntityAttachS2CPacket packet) {

    }

    @Override
    default void onEntityPassengersSet(EntityPassengersSetS2CPacket packet) {

    }

    @Override
    default void onExplosion(ExplosionS2CPacket packet) {

    }

    @Override
    default void onGameStateChange(GameStateChangeS2CPacket packet) {

    }

    @Override
    default void onKeepAlive(KeepAliveS2CPacket packet) {

    }

    @Override
    default void onChunkData(ChunkDataS2CPacket packet) {

    }

    @Override
    default void onUnloadChunk(UnloadChunkS2CPacket packet) {

    }

    @Override
    default void onWorldEvent(WorldEventS2CPacket packet) {

    }

    @Override
    default void onGameJoin(GameJoinS2CPacket packet) {

    }

    @Override
    default void onEntity(EntityS2CPacket packet) {

    }

    @Override
    default void onPlayerPositionLook(PlayerPositionLookS2CPacket packet) {

    }

    @Override
    default void onParticle(ParticleS2CPacket packet) {

    }

    @Override
    default void onPing(PlayPingS2CPacket packet) {

    }

    @Override
    default void onPlayerAbilities(PlayerAbilitiesS2CPacket packet) {

    }

    @Override
    default void onPlayerList(PlayerListS2CPacket packet) {

    }

    @Override
    default void onEntitiesDestroy(EntitiesDestroyS2CPacket packet) {

    }

    @Override
    default void onRemoveEntityStatusEffect(RemoveEntityStatusEffectS2CPacket packet) {

    }

    @Override
    default void onPlayerRespawn(PlayerRespawnS2CPacket packet) {

    }

    @Override
    default void onEntitySetHeadYaw(EntitySetHeadYawS2CPacket packet) {

    }

    @Override
    default void onUpdateSelectedSlot(UpdateSelectedSlotS2CPacket packet) {

    }

    @Override
    default void onScoreboardDisplay(ScoreboardDisplayS2CPacket packet) {

    }

    @Override
    default void onEntityTrackerUpdate(EntityTrackerUpdateS2CPacket packet) {

    }

    @Override
    default void onEntityVelocityUpdate(EntityVelocityUpdateS2CPacket packet) {

    }

    @Override
    default void onEntityEquipmentUpdate(EntityEquipmentUpdateS2CPacket packet) {

    }

    @Override
    default void onExperienceBarUpdate(ExperienceBarUpdateS2CPacket packet) {

    }

    @Override
    default void onHealthUpdate(HealthUpdateS2CPacket packet) {

    }

    @Override
    default void onTeam(TeamS2CPacket packet) {

    }

    @Override
    default void onScoreboardPlayerUpdate(ScoreboardPlayerUpdateS2CPacket packet) {

    }

    @Override
    default void onPlayerSpawnPosition(PlayerSpawnPositionS2CPacket packet) {

    }

    @Override
    default void onWorldTimeUpdate(WorldTimeUpdateS2CPacket packet) {

    }

    @Override
    default void onPlaySound(PlaySoundS2CPacket packet) {

    }

    @Override
    default void onPlaySoundFromEntity(PlaySoundFromEntityS2CPacket packet) {

    }

    @Override
    default void onItemPickupAnimation(ItemPickupAnimationS2CPacket packet) {

    }

    @Override
    default void onEntityPosition(EntityPositionS2CPacket packet) {

    }

    @Override
    default void onEntityAttributes(EntityAttributesS2CPacket packet) {

    }

    @Override
    default void onEntityStatusEffect(EntityStatusEffectS2CPacket packet) {

    }

    @Override
    default void onSynchronizeTags(SynchronizeTagsS2CPacket packet) {

    }

    @Override
    default void onEndCombat(EndCombatS2CPacket packet) {

    }

    @Override
    default void onEnterCombat(EnterCombatS2CPacket packet) {

    }

    @Override
    default void onDeathMessage(DeathMessageS2CPacket packet) {

    }

    @Override
    default void onDifficulty(DifficultyS2CPacket packet) {

    }

    @Override
    default void onSetCameraEntity(SetCameraEntityS2CPacket packet) {

    }

    @Override
    default void onWorldBorderInitialize(WorldBorderInitializeS2CPacket packet) {

    }

    @Override
    default void onWorldBorderInterpolateSize(WorldBorderInterpolateSizeS2CPacket packet) {

    }

    @Override
    default void onWorldBorderSizeChanged(WorldBorderSizeChangedS2CPacket packet) {

    }

    @Override
    default void onWorldBorderWarningTimeChanged(WorldBorderWarningTimeChangedS2CPacket packet) {

    }

    @Override
    default void onWorldBorderWarningBlocksChanged(WorldBorderWarningBlocksChangedS2CPacket packet) {

    }

    @Override
    default void onWorldBorderCenterChanged(WorldBorderCenterChangedS2CPacket packet) {

    }

    @Override
    default void onPlayerListHeader(PlayerListHeaderS2CPacket packet) {

    }

    @Override
    default void onResourcePackSend(ResourcePackSendS2CPacket packet) {

    }

    @Override
    default void onBossBar(BossBarS2CPacket packet) {

    }

    @Override
    default void onCooldownUpdate(CooldownUpdateS2CPacket packet) {

    }

    @Override
    default void onVehicleMove(VehicleMoveS2CPacket packet) {

    }

    @Override
    default void onAdvancements(AdvancementUpdateS2CPacket packet) {

    }

    @Override
    default void onSelectAdvancementTab(SelectAdvancementTabS2CPacket packet) {

    }

    @Override
    default void onCraftFailedResponse(CraftFailedResponseS2CPacket packet) {

    }

    @Override
    default void onCommandTree(CommandTreeS2CPacket packet) {

    }

    @Override
    default void onStopSound(StopSoundS2CPacket packet) {

    }

    @Override
    default void onCommandSuggestions(CommandSuggestionsS2CPacket packet) {

    }

    @Override
    default void onSynchronizeRecipes(SynchronizeRecipesS2CPacket packet) {

    }

    @Override
    default void onLookAt(LookAtS2CPacket packet) {

    }

    @Override
    default void onNbtQueryResponse(NbtQueryResponseS2CPacket packet) {

    }

    @Override
    default void onLightUpdate(LightUpdateS2CPacket packet) {

    }

    @Override
    default void onOpenWrittenBook(OpenWrittenBookS2CPacket packet) {

    }

    @Override
    default void onOpenScreen(OpenScreenS2CPacket packet) {

    }

    @Override
    default void onSetTradeOffers(SetTradeOffersS2CPacket packet) {

    }

    @Override
    default void onChunkLoadDistance(ChunkLoadDistanceS2CPacket packet) {

    }

    @Override
    default void onSimulationDistance(SimulationDistanceS2CPacket packet) {

    }

    @Override
    default void onChunkRenderDistanceCenter(ChunkRenderDistanceCenterS2CPacket packet) {

    }

    @Override
    default void onPlayerActionResponse(PlayerActionResponseS2CPacket packet) {

    }

    @Override
    default void onOverlayMessage(OverlayMessageS2CPacket packet) {

    }

    @Override
    default void onSubtitle(SubtitleS2CPacket packet) {

    }

    @Override
    default void onTitle(TitleS2CPacket packet) {

    }

    @Override
    default void onTitleFade(TitleFadeS2CPacket packet) {

    }

    @Override
    default void onTitleClear(ClearTitleS2CPacket packet) {

    }

    @Override
    default void onServerMetadata(ServerMetadataS2CPacket packet) {

    }

    @Override
    default void onChatSuggestions(ChatSuggestionsS2CPacket packet) {

    }

    @Override
    default void onDisconnected(Text reason) {

    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/network/S2CPlayPacketListener.kt
package io.github.svegon.capi.event.network

import io.github.svegon.capi.event.ListenerCollectionFactory
import io.github.svegon.capi.event.ListenerSet
import io.github.svegon.capi.event.network.S2CPlayPacketListener
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.network.listener.ClientPlayPacketListener
import net.minecraft.network.packet.Packet
import net.minecraft.network.packet.s2c.common.*
import net.minecraft.network.packet.s2c.play.*
import net.minecraft.network.packet.s2c.query.PingResultS2CPacket
import net.minecraft.text.Text
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.util.function.Function

interface S2CPlayPacketListener : ClientPlayPacketListener {
    fun apply(packet: Packet<ClientPlayPacketListener?>, info: CallbackInfo?) {
        packet.apply(this)
    }

    override fun isConnectionOpen(): Boolean {
        return false
    }

    override fun onDamageTilt(packet: DamageTiltS2CPacket) {
    }

    override fun onProfilelessChatMessage(packet: ProfilelessChatMessageS2CPacket) {
    }

    override fun onRemoveMessage(packet: RemoveMessageS2CPacket) {
    }

    override fun onChunkBiomeData(packet: ChunkBiomeDataS2CPacket) {
    }

    override fun onPlayerRemove(packet: PlayerRemoveS2CPacket) {
    }

    override fun onBundle(packet: BundleS2CPacket) {
    }

    override fun onEntityDamage(packet: EntityDamageS2CPacket) {
    }

    override fun onEntitySpawn(packet: EntitySpawnS2CPacket) {
    }

    override fun onExperienceOrbSpawn(packet: ExperienceOrbSpawnS2CPacket) {
    }

    override fun onScoreboardObjectiveUpdate(packet: ScoreboardObjectiveUpdateS2CPacket) {
    }

    override fun onEntityAnimation(packet: EntityAnimationS2CPacket) {
    }

    override fun onStatistics(packet: StatisticsS2CPacket) {
    }

    override fun onUnlockRecipes(packet: UnlockRecipesS2CPacket?) {
    }

    override fun onBlockBreakingProgress(packet: BlockBreakingProgressS2CPacket) {
    }

    override fun onSignEditorOpen(packet: SignEditorOpenS2CPacket) {
    }

    override fun onBlockEntityUpdate(packet: BlockEntityUpdateS2CPacket) {
    }

    override fun onBlockEvent(packet: BlockEventS2CPacket) {
    }

    override fun onBlockUpdate(packet: BlockUpdateS2CPacket) {
    }

    override fun onGameMessage(packet: GameMessageS2CPacket) {
    }

    override fun onChatMessage(packet: ChatMessageS2CPacket) {
    }

    override fun onChunkDeltaUpdate(packet: ChunkDeltaUpdateS2CPacket) {
    }

    override fun onMapUpdate(packet: MapUpdateS2CPacket) {
    }

    override fun onCloseScreen(packet: CloseScreenS2CPacket) {
    }

    override fun onInventory(packet: InventoryS2CPacket) {
    }

    override fun onOpenHorseScreen(packet: OpenHorseScreenS2CPacket) {
    }

    override fun onScreenHandlerPropertyUpdate(packet: ScreenHandlerPropertyUpdateS2CPacket) {
    }

    override fun onScreenHandlerSlotUpdate(packet: ScreenHandlerSlotUpdateS2CPacket) {
    }

    override fun onCustomPayload(packet: CustomPayloadS2CPacket) {
    }

    override fun onDisconnect(packet: DisconnectS2CPacket) {
    }

    override fun onEntityStatus(packet: EntityStatusS2CPacket) {
    }

    override fun onEntityAttach(packet: EntityAttachS2CPacket) {
    }

    override fun onEntityPassengersSet(packet: EntityPassengersSetS2CPacket) {
    }

    override fun onExplosion(packet: ExplosionS2CPacket) {
    }

    override fun onGameStateChange(packet: GameStateChangeS2CPacket) {
    }

    override fun onKeepAlive(packet: KeepAliveS2CPacket) {
    }

    override fun onChunkData(packet: ChunkDataS2CPacket) {
    }

    override fun onUnloadChunk(packet: UnloadChunkS2CPacket) {
    }

    override fun onWorldEvent(packet: WorldEventS2CPacket) {
    }

    override fun onGameJoin(packet: GameJoinS2CPacket) {
    }

    override fun onEntity(packet: EntityS2CPacket) {
    }

    override fun onPlayerPositionLook(packet: PlayerPositionLookS2CPacket) {
    }

    override fun onParticle(packet: ParticleS2CPacket) {
    }

    override fun onPlayerAbilities(packet: PlayerAbilitiesS2CPacket) {
    }

    override fun onPlayerList(packet: PlayerListS2CPacket) {
    }

    override fun onEntitiesDestroy(packet: EntitiesDestroyS2CPacket) {
    }

    override fun onRemoveEntityStatusEffect(packet: RemoveEntityStatusEffectS2CPacket) {
    }

    override fun onPlayerRespawn(packet: PlayerRespawnS2CPacket) {
    }

    override fun onEntitySetHeadYaw(packet: EntitySetHeadYawS2CPacket) {
    }

    override fun onUpdateSelectedSlot(packet: UpdateSelectedSlotS2CPacket) {
    }

    override fun onScoreboardDisplay(packet: ScoreboardDisplayS2CPacket) {
    }

    override fun onEntityTrackerUpdate(packet: EntityTrackerUpdateS2CPacket) {
    }

    override fun onEntityVelocityUpdate(packet: EntityVelocityUpdateS2CPacket) {
    }

    override fun onEntityEquipmentUpdate(packet: EntityEquipmentUpdateS2CPacket) {
    }

    override fun onExperienceBarUpdate(packet: ExperienceBarUpdateS2CPacket) {
    }

    override fun onHealthUpdate(packet: HealthUpdateS2CPacket) {
    }

    override fun onTeam(packet: TeamS2CPacket) {
    }

    fun onScoreboardPlayerUpdate(packet: ScoreboardPlayerUpdateS2CPacket?) {
    }

    override fun onPlayerSpawnPosition(packet: PlayerSpawnPositionS2CPacket) {
    }

    override fun onWorldTimeUpdate(packet: WorldTimeUpdateS2CPacket) {
    }

    override fun onPlaySound(packet: PlaySoundS2CPacket) {
    }

    override fun onPlaySoundFromEntity(packet: PlaySoundFromEntityS2CPacket) {
    }

    override fun onItemPickupAnimation(packet: ItemPickupAnimationS2CPacket) {
    }

    override fun onEntityPosition(packet: EntityPositionS2CPacket) {
    }

    override fun onEntityAttributes(packet: EntityAttributesS2CPacket) {
    }

    override fun onEntityStatusEffect(packet: EntityStatusEffectS2CPacket) {
    }

    override fun onEndCombat(packet: EndCombatS2CPacket) {
    }

    override fun onEnterCombat(packet: EnterCombatS2CPacket) {
    }

    override fun onDeathMessage(packet: DeathMessageS2CPacket) {
    }

    override fun onDifficulty(packet: DifficultyS2CPacket) {
    }

    override fun onSetCameraEntity(packet: SetCameraEntityS2CPacket) {
    }

    override fun onWorldBorderInitialize(packet: WorldBorderInitializeS2CPacket) {
    }

    override fun onWorldBorderInterpolateSize(packet: WorldBorderInterpolateSizeS2CPacket) {
    }

    override fun onWorldBorderSizeChanged(packet: WorldBorderSizeChangedS2CPacket) {
    }

    override fun onWorldBorderWarningTimeChanged(packet: WorldBorderWarningTimeChangedS2CPacket) {
    }

    override fun onWorldBorderWarningBlocksChanged(packet: WorldBorderWarningBlocksChangedS2CPacket) {
    }

    override fun onWorldBorderCenterChanged(packet: WorldBorderCenterChangedS2CPacket) {
    }

    override fun onPlayerListHeader(packet: PlayerListHeaderS2CPacket) {
    }

    override fun onResourcePackSend(packet: ResourcePackSendS2CPacket) {
    }

    override fun onBossBar(packet: BossBarS2CPacket) {
    }

    override fun onCooldownUpdate(packet: CooldownUpdateS2CPacket) {
    }

    override fun onVehicleMove(packet: VehicleMoveS2CPacket) {
    }

    override fun onAdvancements(packet: AdvancementUpdateS2CPacket) {
    }

    override fun onSelectAdvancementTab(packet: SelectAdvancementTabS2CPacket) {
    }

    override fun onCraftFailedResponse(packet: CraftFailedResponseS2CPacket) {
    }

    override fun onCommandTree(packet: CommandTreeS2CPacket) {
    }

    override fun onStopSound(packet: StopSoundS2CPacket) {
    }

    override fun onCommandSuggestions(packet: CommandSuggestionsS2CPacket) {
    }

    override fun onSynchronizeRecipes(packet: SynchronizeRecipesS2CPacket) {
    }

    override fun onLookAt(packet: LookAtS2CPacket) {
    }

    override fun onNbtQueryResponse(packet: NbtQueryResponseS2CPacket) {
    }

    override fun onLightUpdate(packet: LightUpdateS2CPacket) {
    }

    override fun onOpenWrittenBook(packet: OpenWrittenBookS2CPacket) {
    }

    override fun onOpenScreen(packet: OpenScreenS2CPacket) {
    }

    override fun onSetTradeOffers(packet: SetTradeOffersS2CPacket) {
    }

    override fun onChunkLoadDistance(packet: ChunkLoadDistanceS2CPacket) {
    }

    override fun onSimulationDistance(packet: SimulationDistanceS2CPacket) {
    }

    override fun onChunkRenderDistanceCenter(packet: ChunkRenderDistanceCenterS2CPacket) {
    }

    override fun onPlayerActionResponse(packet: PlayerActionResponseS2CPacket) {
    }

    override fun onOverlayMessage(packet: OverlayMessageS2CPacket) {
    }

    override fun onSubtitle(packet: SubtitleS2CPacket) {
    }

    override fun onTitle(packet: TitleS2CPacket) {
    }

    override fun onTitleFade(packet: TitleFadeS2CPacket) {
    }

    override fun onTitleClear(packet: ClearTitleS2CPacket) {
    }

    override fun onServerMetadata(packet: ServerMetadataS2CPacket) {
    }

    override fun onChatSuggestions(packet: ChatSuggestionsS2CPacket) {
    }

    fun onDisconnected(reason: Text?) {
    }

    override fun onPingResult(packet: PingResultS2CPacket) {
    }

    override fun onPing(packet: CommonPingS2CPacket) {
    }

    override fun onSynchronizeTags(packet: SynchronizeTagsS2CPacket) {
    }

    override fun onEnterReconfiguration(packet: EnterReconfigurationS2CPacket) {
    }

    override fun onStartChunkSend(packet: StartChunkSendS2CPacket) {
    }

    override fun onChunkSent(packet: ChunkSentS2CPacket) {
    }

    companion object {
        @JvmField
        val CLIENT_PACKET_RECEIVED_EVENT: Event<S2CPlayPacketListener> = EventFactory.createArrayBacked(
            S2CPlayPacketListener::class.java
        ) { listeners: Array<S2CPlayPacketListener> ->
            object : S2CPlayPacketListener {
                override fun apply(packet: Packet<ClientPlayPacketListener?>, info: CallbackInfo) {
                    for (listener in listeners) {
                        listener.apply(packet, info)

                        if (info.isCancelled) {
                            return@createArrayBacked
                        }
                    }
                }
            }
        }
        val LISTENER_LIST: io.github.svegon.capi.event.ListenerList<S2CPlayPacketListener> =
            ListenerCollectionFactory.Companion.listenersVector<Any>(
                Function<List<Any>, Any> { l: List<Any> ->
                    object : S2CPlayPacketListener {
                        override fun apply(packet: Packet<ClientPlayPacketListener?>, info: CallbackInfo) {
                            for (listener in l) {
                                listener.apply(packet, info)

                                if (info.isCancelled) {
                                    return@listenersVector
                                }
                            }
                        }
                    }
                })
        val LISTENER_SET: ListenerSet<S2CPlayPacketListener> =
            ListenerCollectionFactory.Companion.listenersLinkedSet<Any>(
                Function<Set<Any>, Any> { s: Set<Any> ->
                    object : S2CPlayPacketListener {
                        override fun apply(packet: Packet<ClientPlayPacketListener?>, info: CallbackInfo) {
                            for (listener in s) {
                                listener.apply(packet, info)

                                if (info.isCancelled) {
                                    return@listenersLinkedSet
                                }
                            }
                        }
                    }
                })
    }
}
